Sources/SwiftParser/Names.swift

```diff
@@ -45,6 +45,151 @@ extension Parser {
   }
 }
 
+private enum StructuralTokens: TokenSpecSet {
+  case comma
+  case colon
+  case leftParen
+  case leftBrace
+  case leftSquare
+  case leftAngle
+  case rightParen
+  case rightBrace
+  case rightSquare
+  case rightAngle
+
+  init?(lexeme: Lexer.Lexeme, experimentalFeatures: Parser.ExperimentalFeatures) {
+    switch lexeme.rawTokenKind {
+    case .comma: self = .comma
+    case .colon: self = .colon
+    case .leftParen: self = .leftParen
+    case .leftBrace: self = .leftBrace
+    case .leftSquare: self = .leftSquare
+    case .leftAngle: self = .leftAngle
+    case .rightParen: self = .rightParen
+    case .rightBrace: self = .rightBrace
+    case .rightSquare: self = .rightSquare
+    case .rightAngle: self = .rightAngle
+    default: return nil
+    }
+  }
+
+  var spec: TokenSpec {
+    switch self {
+    case .comma: return .comma
+    case .colon: return .colon
+    case .leftParen: return .leftParen
+    case .leftBrace: return .leftBrace
+    case .leftSquare: return .leftSquare
+    case .leftAngle: return .leftAngle
+    case .rightParen: return .rightParen
+    case .rightBrace: return .rightBrace
+    case .rightSquare: return .rightSquare
+    case .rightAngle: return .rightAngle
+    }
+  }
+}
+
+extension TokenConsumer {
+  /// Do the subsequent tokens have the form of a module selector? Encompasses some invalid syntax which nonetheless
+  /// can be handled by `consumeModuleSelectorTokens()`.
+  ///
+  /// - Postcondition: If `true`, either the current token or the next token is `.colonColon`.
+  mutating func isAtModuleSelector() -> Bool {
+    // If this is a module selector, the next token should be `::`.
+    guard self.peek(isAt: .colonColon) else {
+      // ...however, we will also allow the *current* token to be `::`. `consumeModuleSelectorTokens()` will create a
+      // missing identifier.
+      return self.at(.colonColon)
+    }
+
+    // Technically the current token *should* be an identifier, but we also want to diagnose other tokens that might be
+    // used by accident (particularly keywords and `_`). However, we don't want to consume tokens which would make the
```

Comment:
Would you show me some examples where consuming keywords as a _invalid_ module identifiers improves the diagnostics?

IMO, keyword as a module name rule should follow existing keyword rules should be like:
* Bare keywords as module names are allowed after `.` (following SE-0071). Otherwise it's not a module selector even if there's `::` after that
* Some special keywords such as `Self`, `self` and `Any` should be parsed as _reserved_ names (even without `.`). And the type checker should reject/diagnose them accordingly.

Reply:
I'll narrow this to accept `.wildcard`, `.binaryOperator`, and the keywords `self`, `Self`, `super`, and `Any` in addition to identifiers. Does that sound like a reasonable list?

Reply:
Ah `foo + ::bar` is parsed as a regular binary operator expression either way. Hmm maybe fine then...  🤔